### ISSUE TITLE
move images out of CloudWatt

### ISSUE THREAD

**Issue by maxlath** [inventaire/inventaire#310]
[CloudWatt is shutting down in february 2020](https://support.cloudwatt.com/kb/faq/fermeture/index.html), and that's what inventaire.io uses as an [OpenStack Swift](https://en.wikipedia.org/wiki/OpenStack#Object_storage_(Swift)) infra (see [`server/controllers/images/lib/swift_client.coffee`](https://github.com/inventaire/inventaire/blob/7d29fd3/server/controllers/images/lib/swift_client.coffee))

### Move to another OpenStack Swift provider

**Pros:**
* we would just need to update the config and move the images there, and we are done

**Cons:**
* this exposes us to that provider closing down too at some point, and having to do it all over again?
* images in an OpenStack Swift container aren't so straight forward to manipulate: no bulk API?

**Options**
* [list of providers](https://www.openstack.org/marketplace/public-clouds/)
* [OVH seems like a reasonable option](https://www.ovh.com/fr/public-cloud/prices/#storage)
### Switch to a file system based client

**Pros:**
* [already implemented](https://github.com/inventaire/inventaire/blob/7d29fd3/server/controllers/images/lib/swift_client.coffee) as that's what is used in development
* makes the setup of a production instance easier
* it will never send us a email telling us that it will shut down in 6 month
* probably cheaper

**Cons:**
* the current implementation would require some optimizations (we can't just keep putting hundreds of thousands of files in a single directory, can we?) and a backup system in prod (rsync cron?)
* requires more maintenance and cognitive load
* how far can that scale?

**Comment by maxlath** [comment]
Readings:
* [Why Object Storage?](https://blog.westerndigital.com/why-object-storage/)

**Comment by maxlath** [comment]
we might have the possibility to use [bulk operations](https://www.swiftstack.com/docs/admin/middleware/bulk.html) to move media from one provider to the other

**Comment by maxlath** [comment]
transfer to OVH done \o/

the transfer process went something like this:

## Code changes
Changing provider requires:
* to check for OpenStack Identity version and update [`get_swift_token`](https://github.com/inventaire/inventaire/blob/15bd230/server/controllers/images/lib/get_swift_token.js) accordingly : here there was a change from v2 to v3, addressed in 15bd230c
* to check for OpenStack Swift version and update [`swift` client](https://github.com/inventaire/inventaire/blob/15bd230/server/controllers/images/lib/swift_client.js) accordingly : no change required in this change of provider
* update [OpenStack Swift credentials in configuration](https://github.com/inventaire/inventaire/blob/15bd230/config/default.js#L138-L145): should be done in `./config/local.js` on production server
* restart production server with code changes and configuration update before starting to download/prune/upload, so that newly uploaded images directly go to the new provider containers and don't mess with the pruning process. In the meantime, it will result in 404 errors for images that weren't yet re-uploaded to the new provider, but this remains quite limitied, thank to Nginx caching resized images, and web browsers caching images thank to the [`Cache-control: immutable` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Revalidation_and_reloading).

## Transfer preparation

### Tools
* install [`swift` CLI](https://docs.openstack.org/ocata/cli-reference/swift.html): `sudo apt-get install python-swiftclient`

### Configuration files
Get the configuration files required by swift CLI
* download `openrc.sh` file from CloudWatt and name it `openrc_cloudwatt.sh`
* download `openrc.sh` file from OVH and name it `openrc_ovh.sh` (see [OVH documentation](https://docs.ovh.com/fr/public-cloud/charger-les-variables-denvironnement-openstack/))

## Download
```sh
# Load the environment variables for CloudWatt
source openrc_cloudwatt.sh

download_container(){
  name=$1
  mkdir "$name" && cd "$name" && swift download "$name" && cd ..
}

download_container assets
download_container users
download_container entities
```

## Prune
As there is no process for deleting unused users and entities images, this is a good occasion to get rid of unused ones, which can be done by generating a list of ids from a CouchDB view

### users
```coffee
# http://localhost:5984/users-prod/_design/users_media/_view/used_image_hashes
(doc)->
  # doc.picture with a locally stored image lookes like: /img/users/e3a365f04b263fec9b0096f28a93d276fc35f062
  if doc.picture? && doc.picture && doc.picture.slice(0, 4) isnt 'http'
    emit doc.picture.split('/').slice(-1)[0], null
```
then in a terminal
```sh
# get users used image hashes
curl http://localhost:5984/users-prod/_design/users_media/_view/used_image_hashes > users_used_image_hashes.json
# generate a file with one hash per line
cat users_used_image_hashes.json | jq -cr '.rows | .[].key' > users_used_image_hashes

# move used images in a new folder
mkdir users_used_images
cd users
cat ../users_used_image_hashes | xargs mv -t ../users_used_images
```

### entities
```coffee
# http://localhost:5984/entities-prod/_design/entities_media/_view/used_image_hashes
(doc)->
  if doc.type is 'entity' and doc.claims['invp:P2'] and doc.claims['invp:P2'][0]
    emit doc.claims['invp:P2'][0], null
```
then in a terminal
```sh
# get entities used image hashes
curl http://localhost:5984/entities-prod/_design/entities_media/_view/used_image_hashes > entities_used_image_hashes.json
# generate a file with one hash per line
cat entities_used_image_hashes.json | jq -cr '.rows | .[].key' > entities_used_image_hashes

# move used images in a new folder
mkdir entities_used_images
cd entities
cat ../entities_used_image_hashes | xargs mv -t ../entities_used_images
```

## Upload
In a new terminal session
```sh
# Load the environment variables for OVH
source openrc_ovh.sh

# Create the new public containers
# see https://docs.openstack.org/swift/latest/overview_acl.html
swift post assets --read-acl '.r:*,.rlistings'
swift post users --read-acl '.r:*,.rlistings'
swift post entities --read-acl '.r:*,.rlistings'

# Re-upload used images
cd assets && swift upload assets . && cd ..
cd users_used_images && swift upload users . && cd ..
cd entities_used_images && swift upload entities . && cd ..
```

## Recover upload
The swift tool is quite resilient to network errors, but after a too many failing attemps, some connections will stop to retry, so you might see a drop in upload speed (can be checked using `nethogs`). It might be worth interrupting the process, removing the images already uploaded from the current folder, and restart the upload:
```sh
cd entities_used_images
# get a list of already uploaded images names
swift list entities > ../entities_uploaded_images_hashes
# prepare a directory to store those
mkdir ../entities_uploaded_images
# move them to that new directory
cat ../entities_uploaded_images_hashes | xargs mv -t ../entities_uploaded_images
# restart upload process with only the images that weren't already uploaded
swift upload entities .
```
